typing/typecore.ml

```diff
@@ -670,92 +674,109 @@ end) = struct
         in
         List.find check_type lbls
 
+  (** [disambiguate] selects a concrete description for [lid] using some
+      contextual information: an optional [expected_type], and a list of candidates [lbls].
+      If [expected_type] is [None], it just returns the head of [lbls].
+      Otherwise, it extracts the description directly from the corresponding type definition,
+      except for extension types, where it has to choose it inside [lbls].
```

Comment:
Actually, the code first tries to look up the answer among lbls, so no specific code path is needed for that. The point I was making here is that, for all other types, the answer will be the one found through the type definition, and the change in order is only for warning purposes.

Reply:
I still don't understand what happens for extension types and why that needs to be precised.

Reply:
Just because for extension types there is no information in the type definitions, so the only possible approach is to collect the candidates in scope.
```
except for extension types, for which the information is not available,
so that one has to select it among [lbls] instead.
```

Reply:
What happens when an extension type is passed as expected type to `disambiguate`? The only control path I can see that matches your explanation is that it takes the `Some(tpath, tpath0, principal)` branch, for some reason `principal` is `false`, so we match on `lbls`, we take the first label suggested, it is in fact at the right path, it is not ambiguous, so no warning is shown.

Is that hypothesis correct?

Reply:
(And also: if this reasoning holds, why would extension types satisfy `not principal` in the first place?)

Reply:
I find the location of the comment about extension types a bit misleading: the `disambiguate` has no special logic for them itself. It is the `Constructor` module that does not return any candidates for extension in its `lookup_all_from_type` implementations, and similarly it is `Env.lookup_all_constructors` that filters duplicate extension constructions from its return.

Reply:
I improve on this in #9196 with a better comment and better placement.